<?xml version="1.0"?>
<launch>
   <!--  The file try to save hash table into txt file for further loop closure
     hall3floor
   -->
  <machine name="hawaii" address="192.168.1.3" env-loader="/opt/ros/jade/env.sh" default="true"/>
  <machine name="carmen" address="192.168.1.66" user="mpig" password="2" env-loader="/home/mpig/mpig_catkin_ws/devel/env.sh"/>

  <!--node name="map_server" pkg="map_server" type="map_server" args="$(find carmen_navigation)/maps/hall3floor.yaml"/-->

  <node pkg="tf" type="static_transform_publisher" name="laser_to_base" args="0.0 0.0 0.0 0 0 0.0 /base_link /laser 100" machine="carmen"/>

        <arg name="pi/2" value="1.5707963267948966" />
    <arg name="optical_rotate" value="0 0 0 -$(arg pi/2) 0 -$(arg pi/2)" />
    <node pkg="tf" type="static_transform_publisher" name="camera_to_base"
          args="$(arg optical_rotate) /base_link /stereo_camera 100" machine="carmen"/>

  <!--start hokuyo laser scanner node -->
  <node name="hokuyo" pkg="hokuyo_node" type="hokuyo_node" respawn="false" output="screen" machine="carmen">
      <!-- Set the port to connect to here -->
      <param name="port" type="string" value="/dev/ttyACM0"/>
      <!-- Starts up faster, but timestamps will be inaccurate. --> 
      <param name="calibrate_time" type="bool" value="false"/>         
  </node>

  <!--start carmen driver node -->
  <node name="carmen_driver" pkg="carmen_driver" type="carmen_driver_node" output="screen" machine="carmen">
   <!-- <rosparam file="$(find carmen_driver)/config/driver.yaml" command="load" />  -->
  </node>
     <arg name="dir"  default="file:///home/mpig/mpig_catkin_ws/src/carmen_robot/camera_info"/>    <!-- The directory of carmera parameters     --> 
     <!-- call carmera1394 stereo driver -->
    <node pkg="camera1394stereo" type="camera1394stereo_node" name="camera1394stereo_node" output="screen" machine="carmen">
      <param name="video_mode" value="format7_mode3" />
      <param name="format7_color_coding" value="raw16" />
      <param name="bayer_pattern" value="bggr" />
      <param name="bayer_method" value="" />
      <param name="stereo_method" value="Interlaced" />
      <param name="auto_focus" value="2" />
      <param name="auto_exposure" value="2" />
      <param name="camera_info_url_left" value="$(arg dir)/left.yaml" />
      <param name="camera_info_url_right" value="$(arg dir)/right.yaml" />
      <!--<param name="frame_rate" value="5" />  -->
    </node>

   <!-- Run the ROS package stereo_image_proc for image rectification -->
   <group ns="/stereo_camera" >  
          <node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc" machine="carmen">   
          	<param name="approximate_sync" value="true" />         
          </node>
   </group>

   <!--- Run AMCL -->
   <include file="$(find carmen_navigation)/launch/include/amcl.launch" />


    <!-- Just to uncompress images for loop closure -->
   <node name="republish_left"  type="republish" pkg="image_transport" args="compressed in:=/stereo_camera/left/image_rect  raw out:=/stereo_camera/left/image_rect_uncompressed" />
   <node name="republish_right" type="republish" pkg="image_transport" args="compressed in:=/stereo_camera/right/image_rect raw out:=/stereo_camera/right/image_rect_uncompressed" />

   <node pkg="carmen_navigation" type="hacol_node" name="loop_closure" output="screen">
      <param name="model_file" value="/home/s/initialization_data/trainning_data.xml" />
      <param name="infer" value="true" />
      <param name="lc_out_path" value="/home/s/initialization_data" />
      <param name="camera_topic" value="/stereo_camera" />
      <param name="voc_file" value="/home/s/source/dloop/resources/surf64_k10L6.voc.gz" />
      <param name="learning_img_dir" value="/home/s/initialization_data/images/learning" />
      <param name="img_witdh" value="640" />
      <param name="img_height" value="480" />  
      <remap from="/stereo_camera/left/image_rect" to="/stereo_camera/left/image_rect_uncompressed"/>
      <remap from="/stereo_camera/right/image_rect" to="/stereo_camera/right/image_rect_uncompressed"/>
   </node>

  <!--- Run Move Base -->
  <include file="$(find carmen_navigation)/launch/include/dis_move_base.launch" />
  <include file="$(find carmen_driver)/launch/include/keyboard_tele.launch.xml"/>
  <include file="$(find carmen_driver)/launch/teleop_joystick.launch" />

   <!--node pkg="rviz" type="rviz" name="rviz" args="-d $(find carmen_sim)/launch/rviz/navigation.rviz" /--> 
</launch>
